File: code/scenes/src/main/java/org/fbme/scenes/controllers/components/ComponentsFacility.java
package org.fbme.scenes.controllers.components;

import jetbrains.mps.nodeEditor.cells.EditorCell;
import jetbrains.mps.openapi.editor.EditorContext;
import jetbrains.mps.openapi.editor.cells.CellAction;
import jetbrains.mps.openapi.editor.cells.CellActionType;
import org.fbme.scenes.controllers.*;
import org.fbme.scenes.controllers.scene.*;
import org.fbme.scenes.viewmodel.ComponentsView;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.awt.*;
import java.util.List;
import java.util.*;

public class ComponentsFacility<CompT, FormT> {

    Map<CompT, ComponentEntry<CompT, FormT>> myComponents = new HashMap<>();

    private static final SceneStateKey<Set<?>> SELECTION_KEY = new SceneStateKey<>("comp-selection");

    final ComponentsView<CompT> myView;
    final SceneEditor myEditor;

    final ComponentControllerFactory<CompT, FormT> myControllerFactory;

    final LayoutModel<CompT> myLayout;
    final SelectionModel<CompT> mySelection;

    final ComponentSynchronizer<CompT, FormT> myComponentSyncronizer;
    final SceneFocusModel mySceneFocus;
    final FocusHandle myFocusHandle;

    private final CellAction myDeleteSelectedComponents = new CellAction() {
        @Override
        public String getDescriptionText() {
            return "Delete selected components";
        }

        @Override
        public boolean executeInCommand() {
            return true;
        }

        @Override
        public boolean canExecute(EditorContext context) {
            return true;
        }

        @Override
        public void execute(EditorContext context) {
            for (CompT component : mySelection.getSelectedComponents()) {
                myView.remove(component);
            }
        }
    };

    private boolean myClickShouldSelect = true;

    public ComponentsFacility(
            SceneEditor editor,
            ComponentsView<CompT> componentsView,
            ComponentControllerFactory<CompT, FormT> controllerFactory,
            ComponentSynchronizer<CompT, FormT> componentSyncronizer,
            LayoutModel<CompT> layout,
            SelectionModel<CompT> selection,
            SceneFocusModel sceneFocus,
            Layer componentsLayer,
            Layer tracesLayer
    ) {
        myEditor = editor;
        myView = componentsView;
        myControllerFactory = controllerFactory;
        myComponentSyncronizer = componentSyncronizer;
        myLayout = layout;
        mySelection = selection;
        mySceneFocus = sceneFocus;
        myFocusHandle = new FocusHandle(() -> {
            mySelection.clear();
            clearSelection();
        });

        init();

        editor.addCellActionProvider(new MyCellActionProvider());
        editor.addLayouter(new MyLayouter());
        editor.addCellProvider(componentsLayer, new MyCellProvider());
        editor.addClickListener(componentsLayer, new MyClickEventListener());
        editor.addDragListener(componentsLayer, new MyDragEventListener());
        editor.addPainter(tracesLayer, new MyPainter());
        editor.addInitializer(new MySelectionInitializer());
    }

    private void init() {
        Set<CompT> viewComponents = myView.getComponents();

        for (CompT component : viewComponents) {
            ComponentEntry<CompT, FormT> entry = new ComponentEntry<>(ComponentsFacility.this, component);
            myComponents.put(component, entry);
            myLayout.addComponent(component, entry.myLayoutSetting);
        }
    }

    private void setSelection(@NotNull CompT component, boolean selected) {
        mySelection.setSelected(component, selected);
        if (selected) {
            mySceneFocus.addFocus(myFocusHandle);
        }
    }

    private void clearSelection() {
        for (ComponentEntry<CompT, FormT> component : myComponents.values()) {
            component.myController.updateCellSelection(false);
        }
    }

    @NotNull
    public ComponentController<FormT> getController(@NotNull CompT component) {
        ComponentEntry<CompT, FormT> entry = myComponents.get(component);
        return entry.myController;
    }

    @NotNull
    public FormT getModelForm(@NotNull CompT component) {
        return myComponents.get(component).myModelForm;
    }

    @Nullable
    public FormT getTransformedForm(@NotNull CompT component) {
        return myComponents.get(component).myTransformedForm;
    }

    public ComponentSynchronizer<CompT, FormT> getComponentSyncronizer() {
        return myComponentSyncronizer;
    }

    private class MyLayouter implements SceneLayouter {

        @Override
        public void relayout() {
            for (ComponentEntry<CompT, FormT> component : myComponents.values()) {
                component.relayout();
            }
        }

        @NotNull
        @Override
        public Rectangle getLayoutBounds() {
            Rectangle bounds = new Rectangle(-1, -1);
            for (ComponentEntry<CompT, FormT> component : myComponents.values()) {
                bounds = bounds.union(component.getLayoutBounds());
            }
            return bounds;
        }
    }

    private class MyPainter implements ScenePainter {

        @Override
        public void paint(Graphics2D graphics) {
            for (Map.Entry<CompT, ComponentEntry<CompT, FormT>> component : myComponents.entrySet()) {
                if (myLayout.getTracePosition(component.getKey()) != null) {
                    ComponentEntry<CompT, FormT> entry = component.getValue();
                    entry.myController.paintTrace(graphics.create(), entry.myModelForm);
                }
            }
        }
    }

    private class MyCellProvider implements EditorCellProvider {

        @Override
        public List<EditorCell> getCells() {
            List<EditorCell> cells = new ArrayList<>();
            for (ComponentEntry<CompT, FormT> component : myComponents.values()) {
                EditorCell cell = component.myController.getComponentCell();
                cells.add(cell);
            }
            return cells;
        }
    }

    private class MyCellActionProvider implements SceneCellActionProvider {

        @NotNull
        @Override
        public List<CellActionType> getProvidedActionsTypes() {
            return Collections.singletonList(CellActionType.BACKSPACE);
        }

        @Nullable
        @Override
        public CellAction getAction(@NotNull CellActionType action) {
            if (action == CellActionType.BACKSPACE) {
                return myDeleteSelectedComponents;
            }
            return null;
        }
    }

    private class MySelectionInitializer implements SceneInitializer {

        @Override
        public void onAdd() {
            @SuppressWarnings("unchecked")
            Set<CompT> selection = (Set<CompT>) myEditor.loadState(SELECTION_KEY);
            if (selection != null) {
                for (CompT selected : selection) {
                    setSelection(selected, true);
                }
            }
        }

        @Override
        public void onRemove() {
            myEditor.storeState(SELECTION_KEY, mySelection.getSelectedComponents());
        }
    }

    private class MyClickEventListener implements ClickEventListener {

        @Override
        public void onMouseClicked(ClickEvent event) {
            CompT component = myLayout.findAt(event.awt.getX(), event.awt.getY());
            if (component != null) {
                setSelection(component, myClickShouldSelect);
                myComponents.get(component).relayout();
                event.consume();
            }
        }
    }

    private void syncOnMovement(CompT component, int dx, int dy) {
        ComponentEntry<CompT, FormT> entry = myComponents.get(component);
        entry.myTransformedForm = entry.myController.translateForm(entry.myModelForm, dx, dy);
    }

    private void syncOnMovementCompletion(CompT component) {
        myComponents.get(component).myTransformedForm = null;
    }

    private class MyDragEventListener implements DragEventListener {

        @Override
        public void onDragStarted(DragEvent event) {
            CompT component = myLayout.findAt(event.awt.getX(), event.awt.getY());
            if (component != null) {
                Set<CompT> movingComponents = new HashSet<>();
                movingComponents.add(component);
                if (event.awt.isMetaDown()) {
                    movingComponents.addAll(mySelection.getSelectedComponents());
                } else {
                    mySceneFocus.clearFocus();
                    mySelection.clear();
                }
                myClickShouldSelect = !(mySelection.isSelected(component));
                setSelection(component, true);
                myComponents.get(component).relayout();
                MyDragEventHandler handler = new MyDragEventHandler(movingComponents, myLayout.moveComponents(movingComponents), new Point(event.awt.getX(), event.awt.getY()));
                event.consume(handler);
                myEditor.fireRepaint();
            }
        }
    }

    private class MyDragEventHandler implements DragEventHandler {

        private final LayoutMovememntHandle myHandle;
        private final Point myOrigin;
        private final Set<CompT> myMovedComponents;

        public MyDragEventHandler(Set<CompT> components, LayoutMovememntHandle handle, Point origin) {
            myHandle = handle;
            myOrigin = origin;
            myMovedComponents = components;
        }

        @Override
        public void dragTo(int x, int y) {
            int dx = x - myOrigin.x;
            int dy = y - myOrigin.y;
            myHandle.moveTo(dx, dy);

            myEditor.fireRelayout();

            for (CompT component : myMovedComponents) {
                syncOnMovement(component, dx, dy);
            }
        }

        @Override
        public void completeAt(int x, int y) {
            myHandle.moveTo(x - myOrigin.x, y - myOrigin.y);
            myHandle.complete();

            myEditor.fireRelayout();

            for (CompT component : myMovedComponents) {
                syncOnMovementCompletion(component);
            }
        }
    }
}

File: code/scenes/src/main/kotlin/org/fbme/scenes/controllers/components/ComponentsFacility.kt
package org.fbme.scenes.controllers.components

import jetbrains.mps.nodeEditor.cells.EditorCell
import jetbrains.mps.openapi.editor.EditorContext
import jetbrains.mps.openapi.editor.cells.CellAction
import jetbrains.mps.openapi.editor.cells.CellActionType
import org.fbme.scenes.controllers.*
import org.fbme.scenes.controllers.scene.*
import org.fbme.scenes.viewmodel.ComponentsView
import java.awt.Graphics2D
import java.awt.Point
import java.awt.Rectangle

class ComponentsFacility<CompT, FormT>(
    val editor: SceneEditor,
    val view: ComponentsView<CompT>,
    val controllerFactory: ComponentControllerFactory<CompT, FormT>,
    val componentSynchronizer: ComponentSynchronizer<CompT, FormT>,
    val layout: LayoutModel<CompT>,
    val selection: SelectionModel<CompT>,
    val sceneFocus: SceneFocusModel,
    componentsLayer: Layer,
    tracesLayer: Layer
) {
    var components: MutableMap<CompT, ComponentEntry<CompT, FormT>> = HashMap()

    private val focusHandle: FocusHandle

    private val myDeleteSelectedComponents: CellAction = object : CellAction {
        override fun getDescriptionText(): String {
            return "Delete selected components"
        }

        override fun executeInCommand(): Boolean {
            return true
        }

        override fun canExecute(context: EditorContext): Boolean {
            return true
        }

        override fun execute(context: EditorContext) {
            for (component in selection.selectedComponents) {
                view.remove(component)
            }
        }
    }

    private var clickShouldSelect = true

    private fun init() {
        val viewComponents = view.components
        for (component in viewComponents) {
            val entry = ComponentEntry(this@ComponentsFacility, component)
            components[component] = entry
            layout.addComponent(component, entry.layoutSetting)
        }
    }

    private fun setSelection(component: CompT, selected: Boolean) {
        selection.setSelected(component, selected)
        if (selected) {
            sceneFocus.addFocus(focusHandle)
        }
    }

    private fun clearSelection() {
        for (component in components.values) {
            component.controller.updateCellSelection(false)
        }
    }

    fun getController(component: CompT): ComponentController<FormT> {
        val entry = components[component]!!
        return entry.controller
    }

    fun getModelForm(component: CompT): FormT {
        return components[component]!!.modelForm
    }

    fun getTransformedForm(component: CompT): FormT? {
        return components[component]!!.transformedForm
    }

    private inner class MyLayouter : SceneLayouter {
        override fun relayout() {
            for (component in components.values) {
                component.relayout()
            }
        }

        override val layoutBounds: Rectangle
            get() {
                var bounds = Rectangle(-1, -1)
                for (component in components.values) {
                    bounds = bounds.union(component.layoutBounds)
                }
                return bounds
            }
    }

    private inner class MyPainter : ScenePainter {
        override fun paint(graphics: Graphics2D) {
            for ((key, entry) in components) {
                if (layout.getTracePosition(key) != null) {
                    entry.controller.paintTrace(graphics.create(), entry.modelForm)
                }
            }
        }
    }

    private inner class MyCellProvider : EditorCellProvider {
        override val cells: List<EditorCell>
            get() {
                val cells: MutableList<EditorCell> = ArrayList()
                for (component in components.values) {
                    val cell = component.controller.componentCell
                    cells.add(cell)
                }
                return cells
            }
    }

    private inner class MyCellActionProvider : SceneCellActionProvider {
        override val providedActionsTypes: List<CellActionType>
            get() = listOf(CellActionType.BACKSPACE)

        override fun getAction(action: CellActionType): CellAction? {
            return if (action == CellActionType.BACKSPACE) {
                myDeleteSelectedComponents
            } else null
        }
    }

    private inner class MySelectionInitializer : SceneInitializer {
        override fun onAdd() {
            val selection = editor.loadState(SELECTION_KEY) as? Set<CompT>
            if (selection != null) {
                for (selected in selection) {
                    setSelection(selected, true)
                }
            }
        }

        override fun onRemove() {
            editor.storeState(SELECTION_KEY, selection.selectedComponents)
        }
    }

    private inner class MyClickEventListener : ClickEventListener {
        override fun onMouseClicked(event: ClickEvent) {
            val component = layout.findAt(event.awt.x, event.awt.y)
            if (component != null) {
                setSelection(component, clickShouldSelect)
                components[component]!!.relayout()
                event.consume()
            }
        }
    }

    private fun syncOnMovement(component: CompT, dx: Int, dy: Int) {
        val entry = components[component]!!
        entry.transformedForm = entry.controller.translateForm(entry.modelForm, dx, dy)
    }

    private fun syncOnMovementCompletion(component: CompT) {
        components[component]!!.transformedForm = null
    }

    private inner class MyDragEventListener : DragEventListener {
        override fun onDragStarted(event: DragEvent) {
            val component = layout.findAt(event.awt.x, event.awt.y)
            if (component != null) {
                val movingComponents: MutableSet<CompT> = HashSet()
                movingComponents.add(component)
                if (MouseEvents.isMetaDown(event.awt)) {
                    movingComponents.addAll(selection.selectedComponents)
                } else {
                    sceneFocus.clearFocus()
                    selection.clear()
                }
                clickShouldSelect = !selection.isSelected(component)
                setSelection(component, true)
                components[component]!!.relayout()
                val handler = MyDragEventHandler(
                    movingComponents,
                    layout.moveComponents(movingComponents),
                    Point(event.awt.x, event.awt.y)
                )
                event.consume(handler)
                editor.fireRepaint()
            }
        }
    }

    private inner class MyDragEventHandler(
        private val myMovedComponents: Set<CompT>,
        private val myHandle: LayoutMovementHandle,
        private val myOrigin: Point
    ) : DragEventHandler {
        override fun dragTo(x: Int, y: Int) {
            val dx = x - myOrigin.x
            val dy = y - myOrigin.y
            myHandle.moveTo(dx, dy)
            editor.fireRelayout()
            for (component in myMovedComponents) {
                syncOnMovement(component, dx, dy)
            }
        }

        override fun completeAt(x: Int, y: Int) {
            myHandle.moveTo(x - myOrigin.x, y - myOrigin.y)
            myHandle.complete()
            editor.fireRelayout()
            for (component in myMovedComponents) {
                syncOnMovementCompletion(component)
            }
        }
    }

    companion object {
        private val SELECTION_KEY = SceneStateKey<Set<*>?>("comp-selection")
    }

    init {
        focusHandle = FocusHandle {
            selection.clear()
            clearSelection()
        }
        init()
        editor.addCellActionProvider(MyCellActionProvider())
        editor.addLayouter(MyLayouter())
        editor.addCellProvider(componentsLayer, MyCellProvider())
        editor.addClickListener(componentsLayer, MyClickEventListener())
        editor.addDragListener(componentsLayer, MyDragEventListener())
        editor.addPainter(tracesLayer, MyPainter())
        editor.addInitializer(MySelectionInitializer())
    }
}
